### PR TITLE
[dapp-kit] use Zustand to model wallet state + handle autoconnect functionality

### PR DESCRIPTION
**Pull request by williamrobertson13**
## Description 

Okay, this is a stab at re-working https://github.com/MystenLabs/sui/pull/13719 using Zustand. By using Zustand for state management, we can A). handle state persistence more elegantly for auto-connection functionality and B). reduce the amount of unnecessary re-renders by using selectors. 

Why Zustand over Jotai / insert other library?
=> ~~I like bears~~ Zustand is a liiiiiittle bit smaller, can be used outside of React if we want to go cross-platform, and was a little easier for me to think in coming from the Redux world. I don't think there's one better or the other, and Jotai could easily be swapped in here. Although Zustand doesn't handle computed properties well after getting into the weeds with this... maybe Jotai would have been better here after all 🤔 

Why slices over creating multiple stores (e.g. store = [walletSlice, clientSlice] vs. stores = [walletStore, clientStore])?
=> Zustand is un-opinionated, and I don't really think this makes any difference at this point. Using slices allows us to derive data where as stores are completely isolated. This was just the first setup I went with (https://docs.pmnd.rs/zustand/guides/slices-pattern)

What about `SuiClientProvider`?
=> I tried to port that over to use Zustand with its own `suiClientSlice`... but the network state is kind of tricky to handle since it can be controlled or uncontrolled 🤔. If we think this is the right approach, then I can production-ize this PR some more tomorrow with that included.

## Test Plan 
- Automated tests pass

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: sdk/dapp-kit/src/hooks/wallet/useConnectionStatus.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useDAppKitStore } from '../useDAppKitStore.js';
+
+/**
+ * Retrieves the
+ */
+export function useConnectionStatus() {
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/hooks/wallet/useConnectionStatus.ts:9]: Ignore the missing comment lol, will fix that

File: sdk/dapp-kit/src/slices/walletSlice.ts [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+		unregisteredWallet: Wallet,
+	) => void;
+};
+
+export type WalletSlice = {
+	wallets: WalletWithRequiredFeatures[];
+	currentAccount: WalletAccount | null;
+	lastConnectedAccountAddress: string | null;
+} & WalletActions &
+	(
```
Comment (williamrobertson13) [inline sdk/dapp-kit/src/slices/walletSlice.ts:26]: I'm not sure if we want the `connectionStatus` field at all, if we want disconnected/connecting/connected as possible options, or just disconnected/connected 🤔 

My arguing point for all three was that you might want this data outside the context of connecting/disconnecting in your application even though it isn't a popular use case

File: sdk/dapp-kit/src/hooks/useDAppKitStore.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { DAppKitContext } from 'dapp-kit/src/contexts/dAppKitContext';
```
Comment (Jordan-Mysten) [inline sdk/dapp-kit/src/hooks/useDAppKitStore.ts:4]: These import paths look wrong

File: sdk/dapp-kit/src/utils/walletUtils.ts [hunk 4]
```diff
@@ -2,18 +2,19 @@
...
-import type { StorageAdapter } from './storageAdapters.js';
+import { getWallets, isWalletWithRequiredFeatureSet } from '@mysten/wallet-standard';
 
-export function sortWallets<AdditionalFeatures extends Wallet['features']>(
-	wallets: readonly Wallet[],
+export function getRegisteredWallets<AdditionalFeatures extends Wallet['features']>(
 	preferredWallets: string[],
 	requiredFeatures?: (keyof AdditionalFeatures)[],
 ) {
+	const walletsApi = getWallets();
```
Comment (Jordan-Mysten) [inline sdk/dapp-kit/src/utils/walletUtils.ts:15]: Is there any downside for having this re-initialized multiple times? I'd maybe dive into the implementation and just verify that there's not (I think we already call it multiple times today as well, I just want to verify before we double down).

File: sdk/dapp-kit/src/hooks/wallet/useAutoConnectWallet.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+import { useEffect } from 'react';
+import { useDAppKitStore } from '../useDAppKitStore.js';
+import { useConnectWallet } from './useConnectWallet.js';
+import { useWallets } from './useWallets.js';
+
+export function useAutoConnectWallet(autoConnectEnabled: boolean) {
+	const { mutate: connectWallet } = useConnectWallet();
+	const wallets = useWallets();
+	const lastConnectedWalletName = useDAppKitStore((state) => state.lastConnectedWalletName);
```
Comment (Jordan-Mysten) [inline sdk/dapp-kit/src/hooks/wallet/useAutoConnectWallet.ts:12]: Fun way of doing this

File: sdk/dapp-kit/src/slices/walletSlice.ts (hunk 2 above)
Comment (Jordan-Mysten) [reply sdk/dapp-kit/src/slices/walletSlice.ts:26]: Yeah I go back and forth, I'm fine with either tbh, it's more that I'm not sure what the value is in practice.

**Review by Jordan-Mysten — APPROVED** [review]

File: sdk/dapp-kit/src/utils/walletUtils.ts (hunk 4 above)
Comment (williamrobertson13) [reply sdk/dapp-kit/src/utils/walletUtils.ts:15]: Nope, if `wallets` has already been initialized  then it's just returns the frozen object

image

**Pull request merged**